esx_service/utils/auth_api.py

```diff
@@ -777,14 +857,29 @@ def check_privilege_parameters(privilege):
     # If both volume max size and volume total size are set,
     # volume max size should not exceed volume total size
     if (volume_maxsize and volume_totalsize and (volume_maxsize > volume_totalsize)):
-        error_info = error_code.generate_error_info(ErrorCode.PRIVILEGE_INVALID_VOLUME_SIZE, volume_maxsize, volume_totalsize)
+        error_info = generate_error_info(ErrorCode.PRIVILEGE_INVALID_VOLUME_SIZE, volume_maxsize, volume_totalsize)
         return error_info
 
     return None
 
+def check_usage_quota(datastore, volume_totalsize_in_MB):
+    """
+        Check the usage_quota to be set in datastore is valid or not
+        Return None if the usage_quota is valid
+        Return error_info if the usage_quota is invalid
+    """
+    # usage_quota on "__VM_DS" and "__ALL_DS" should be unlimited (volume_totalsize_in_MB shoule be 0)
+    if datastore == auth_data_const.VM_DS or datastore == auth_data_const.ALL_DS:
+        if volume_totalsize_in_MB is not None and volume_totalsize_in_MB > 0:
+            err_msg = "Only unlimit is allowed"
```

Comment:
I think "volume_totalsize" only make sense on a specific physical datastore, it does not make sense to have "volume_totalsize" limit on a "_VM_DS" or "_ALL_DS". In addition, when the volume is created, it must be created in a physical datastore, so any row in the "volumes table", the key is <tenant_uuid, datastore_url>, here the "datastore_url" must be a url for a physical datastore.